Unable to resume cached task with enum in parameters to process

## Bug report 

(Please follow this template replacing the text between parentheses with the requested information)

### Expected behavior and actual behavior

When running the following pipeline we would expect the resume mechanism to work without recomputing the values from the process.

### Steps to reproduce the problem

In `./lib/MyEnum.groovy` file
```
public enum MyEnum {
    A,
    B
}
```

in `main.nf` file
```
workflow {
    ch = Channel.of([MyEnum.A, MyEnum.B])

    SomeTask(ch)

}

process SomeTask {
    input:
        tuple val(meta), val(myEnum)
    
    output:
        tuple val(meta), path('out.txt')

    script:
    """
    echo "hello $myEnum" > out.txt
    """
}
```

Running:
```
nextflow run main.nf
nextflow run main.nf -resume
```


### Program output 

```
Mar-14 15:01:45.856 [Actor Thread 3] WARN  nextflow.processor.TaskProcessor - [SomeTask (1)] Unable to resume cached task -- See log file for details
com.esotericsoftware.kryo.KryoException: Unable to find class: MyEnum
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:138)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:115)
        at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:641)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:752)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:143)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:21)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:143)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:21)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:761)
        at com.esotericsoftware.kryo.Kryo$readClassAndObject$5.call(Unknown Source)
        at nextflow.util.KryoHelper.deserialize(SerializationHelper.groovy:181)
        at nextflow.util.KryoHelper.deserialize(SerializationHelper.groovy)
        at nextflow.util.KryoHelper$deserialize$0.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
        at nextflow.processor.TaskContext.deserialize(TaskContext.groovy:202)
        at nextflow.cache.CacheDB.getTaskEntry(CacheDB.groovy:88)
        at nextflow.processor.TaskProcessor.checkCachedOrLaunchTask(TaskProcessor.groovy:791)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:48)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:189)
        at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:57)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203)
        at nextflow.processor.TaskProcessor.invokeTask(TaskProcessor.groovy:639)
        at nextflow.processor.InvokeTaskAdapter.call(InvokeTaskAdapter.groovy:52)
        at groovyx.gpars.dataflow.operator.DataflowOperatorActor.startTask(DataflowOperatorActor.java:120)
        at groovyx.gpars.dataflow.operator.ForkingDataflowOperatorActor.access$001(ForkingDataflowOperatorActor.java:35)
        at groovyx.gpars.dataflow.operator.ForkingDataflowOperatorActor$1.run(ForkingDataflowOperatorActor.java:58)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.ClassNotFoundException: MyEnum
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:467)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.readName(DefaultClassResolver.java:136)
        ... 37 common frames omitted
```

### Environment 

* Nextflow version: 23.10.1 build 5891
* Java version: openjdk version "17.0.10" 2024-01-16 LTS
* Operating system: macOS
* Bash version: (use the command `$SHELL --version`)

### Additional context

--- a/modules/nextflow/src/main/groovy/nextflow/processor/TaskContext.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/processor/TaskContext.groovy
@@ -206,7 +206,9 @@ class TaskContext implements Map<String,Object>, Cloneable {
     }
 
     static TaskContext deserialize(TaskProcessor processor, byte[] buffer) {
-        def map = (Map)KryoHelper.deserialize(buffer)
+        // use the script class loader so that Kryo can resolve user-defined types
+        final loader = processor.ownerScript?.getClass()?.getClassLoader()
+        final map = (Map)KryoHelper.deserialize(buffer, loader)
         new TaskContext(processor, map)
     }
 

--- a/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/util/SerializationHelper.groovy
@@ -182,7 +182,7 @@ class KryoHelper {
         }
         finally {
             if( prev ) {
-                kryo.setClassLoader(loader)
+                kryo.setClassLoader(prev)
             }
         }
     }

--- a/tests/checks/enum.nf/.checks
+++ b/tests/checks/enum.nf/.checks
@@ -1,0 +1,20 @@
+#
+# Verify that a task whose input/output is a user-defined enum declared
+# *inline* in the pipeline script can be resumed from the cache.
+# Regression test for https://github.com/nextflow-io/nextflow/issues/4816
+#
+set -e
+
+#
+# run normal mode
+#
+$NXF_RUN | tee .stdout
+
+[[ `grep INFO .nextflow.log | grep -c 'Submitted process'` == 1 ]] || false
+
+#
+# run resume mode -- the task must be served from the cache
+#
+$NXF_RUN -resume | tee .stdout
+
+[[ `grep INFO .nextflow.log | grep -c 'Cached process'` == 1 ]] || false

--- a/tests/enum.nf
+++ b/tests/enum.nf
@@ -1,0 +1,23 @@
+
+enum MyEnum {
+    A,
+    B
+}
+
+workflow {
+    ch = channel.of( tuple([:], MyEnum.B) )
+    SomeTask(ch)
+}
+
+process SomeTask {
+    input:
+    tuple val(meta), val(myEnum)
+
+    output:
+    tuple val(meta), path('out.txt')
+
+    script:
+    """
+    echo "hello $myEnum" > out.txt
+    """
+}

